### PR TITLE
added cached tests

### PR DESCRIPTION
**Pull request by aaron-prindle**

### CODE REVIEW

**Comment by aaron-prindle** [comment]
cache test uploads same image twice, testing that layers are the same with the exception of a single app layer which differs on each upload (all dep layers should be the same).  Currently the cloudbuild step is added as a separate run, similar to the benchmark test and not included as part of the integration tests but this would be easy to change.

File: ftl/node/cached/main.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
+# Copyright 2017 Google Inc. All rights reserved.
```
Comment (tejal29) [inline ftl/node/cached/main.py]: 2018

File: ftl/cached/cached_test.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
+# Copyright 2017 Google Inc. All Rights Reserved.
```
Comment (tejal29) [inline ftl/cached/cached_test.py]: 2018

File: ftl/cached/args.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
+# Copyright 2017 Google Inc. All Rights Reserved.
```
Comment (tejal29) [inline ftl/cached/args.py]: 2018

File: ftl/cached/cached.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
+# Copyright 2017 Google Inc. All rights reserved.
```
Comment (tejal29) [inline ftl/cached/cached.py]: 2018

File: ftl/cached/ftl_node_cached_yaml.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+                ]
+            },
+            # Get image digests so that we can delete uploaded images
+            {
+                'name':
+                'gcr.io/cloud-builders/gcloud',
+                'entrypoint':
+                '/bin/bash',
+                'args': [
+                    '-c', 'gcloud container images list-tags ' + path + ' | ' +
```
Comment (dlorenc) [inline ftl/cached/ftl_node_cached_yaml.py]: Can you do this part in the test itself?

File: ftl/node/cached/BUILD [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    main = "main.py",
+    deps = [
+        "//ftl:ftl_lib",
+        "//ftl/cached:cached_lib",
+    ],
+)
+
+load("@base_images_docker//dockerfile_build:dockerfile_build.bzl", "dockerfile_build")
+
+dockerfile_build(
```
Comment (dlorenc) [inline ftl/node/cached/BUILD]: Why do we need a dockerfile build here?

File: ftl/cached/cached.py [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+                logging.info('output of build {0}: {1}'.format(label, output))
+                lyr_shas.append(self._fetch_lyr_shas(img_name))
+                self._del_img_from_gcr(img_name)
+            try:
+                self._compare_layers(lyr_shas[0], lyr_shas[1])
+            except RuntimeError as e:
+                logging.error(e)
+                exit(1)
+        except OSError:
+            raise OSError("""Cached test assumes either ftl/{0}_builder.par
```
Comment (dlorenc) [inline ftl/cached/cached.py]: This shouldn't be an oserror. You're also swallowing the real error here, which might be anything.

File: ftl/cached/ftl_python_cached_yaml.py [hunk 8]
```diff
@@ -0,0 +1,72 @@
+"""A script to generate a cloudbuild yaml."""
+
+import os
+import yaml
+
+_TEST_DIR = '/workspace/ftl/python/testdata'
```
Comment (dlorenc) [inline ftl/cached/ftl_python_cached_yaml.py]: Could you refactor the three new cache .py files into one script that takes the language as a parameter?

File: ftl/node/cached/BUILD (hunk 6 above)
Comment (aaron-prindle) [reply ftl/node/cached/BUILD]: removed

File: ftl/cached/cached.py (hunk 7 above)
Comment (aaron-prindle) [reply ftl/cached/cached.py]: removed

File: ftl/cached/ftl_python_cached_yaml.py (hunk 8 above)
Comment (aaron-prindle) [reply ftl/cached/ftl_python_cached_yaml.py]: done

File: ftl/cached/ftl_node_cached_yaml.py (hunk 5 above)
Comment (aaron-prindle) [reply ftl/cached/ftl_node_cached_yaml.py]: done

File: ftl/cached/Dockerfile [hunk 9]
```diff
@@ -0,0 +1,9 @@
+FROM bazel/reproducible:ubuntu_vanilla
```
Comment (dlorenc) [inline ftl/cached/Dockerfile]: Why do we need this Dockerfile?

Comment (aaron-prindle) [reply ftl/cached/Dockerfile]: we don't, i will remove it in another PR as kokoro is a bit flaky rn

**Pull request merged**